src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedAssembliesAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    public sealed class BannedAssembliesAnalyzer : SymbolAnalyzerBase
+    {
+        private static HashSet<string> BannedAssemblies = new HashSet<string>()
+        {
+            "System.Text.Json",
+            "Newtonsoft.Json",
+            "System.Collections.Immutable"
+        };
+
+        private static readonly string BannedAssembliesMessageArgs = string.Join(", ", BannedAssemblies);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Descriptors.AZC0014);
+
+        public override SymbolKind[] SymbolKinds { get; } = new[]
+        {
+            SymbolKind.Method,
+            SymbolKind.Field,
+            SymbolKind.Property,
+            SymbolKind.Parameter,
+            SymbolKind.Event,
+            SymbolKind.NamedType
+        };
+
+        public override void Analyze(ISymbolAnalysisContext context)
+        {
+            void CheckType(ITypeSymbol type, ISymbol symbol)
```

Comment:
Can you make this static so we don't capture `context`?

Reply:
What's wrong with capturing context here?

Reply:
`Analyze` is called for all type members (public and private), so closures may create memory overhead, and local functions don't have publicly defined rules on when closure is generated as a struct or as a class.

Reply:
Even if closure is generated as a class it won't be allocated until the CheckType call that is only done for public APIs.

There is also https://github.com/dotnet/roslyn/blob/master/docs/compilers/Design/Closure%20Conversion.md that describes some rules.

I also checked and this usage does indeed generate struct closure.

Reply:
> Even if closure is generated as a class it won't be allocated until the CheckType call

No, it is allocated [at the beginning of the method](https://sharplab.io/#v2:D4AQDABCCMAsDcBYAUOK0B0AZAlgOwEclkUQBmCfAFwFMAnAMwEMBjGiASWhQG8UIBUWBADCTADbiAFNUoBKYgF8UpAEwQAKjQDOVXv0EAHOjgBuTWkIgBZaFIBGAe0fiI9gDSU8VCACtPMGQAPNQAfBAsnlwQjva+chAGAnzIgmlWWHayOAkAvOGxvhhikjIKSekVaTgMEFIAhPYJVYIp6e1QAOzEHRDKJKkdDI50NKwAFnXmdJReERgA6uP0NDIQ+bPh8c2D7W29gpllPR39af2KQA).

While current implementation generates struct, it is easy to accidentally change that in the future. Static local functions have been introduced specifically to avoid such possibility

Reply:
Good point!